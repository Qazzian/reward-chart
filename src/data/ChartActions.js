import {loadData, saveData} from './LocalStorage';

export const
	CHART_LOAD_REQUEST = 'CHART_LOAD_REQUEST',
	CHART_LOAD_RECEIVE = 'CHART_LOAD_RECEIVE',
	CHART_SAVE_REQUEST = 'CHART_SAVE_REQUEST',
	CHART_SAVE_COMPLETE = 'CHART_SAVE_COMPLETE',
	CHART_ADD = 'CHART_ADD',
	CHART_REMOVE = 'CHART_REMOVE',
	CHART_HAPPY = 'CHART_HAPPY',
	CHART_SAD = 'CHART_SAD';

export function requestChartLoad() {
	return {
		type: CHART_LOAD_REQUEST,
	}
}

export function receiveChartLoad(chartState) {
	return {
		type: CHART_LOAD_RECEIVE,
		data: chartState
	}
}

export function requestChartSave() {
	return {
		type: 'CHART_SAVE_REQUEST'
	}
}

export function completeChartSave() {
	return {
		type: 'CHART_SAVE_COMPLETE'
	}
}

export function addChart(chartData) {
	return {
		type: CHART_ADD,
		data: chartData,
	};
}

export function removeChart(chart) {
	return {
		type: CHART_REMOVE,
		data: chart,
	};
}

export function chartHappy(chart, date) {
	return {
		type: CHART_HAPPY,
		data: chart,
		date: date,
	};
}

export function chartSad(chart, date) {
	return {
		type: CHART_SAD,
		data: chart,
		date: date,
	};
}

export function fetchChartData() {
	return dispatch => {
		dispatch(requestChartLoad());
		return loadData()
			.then(response => dispatch(receiveChartLoad(response)));
	}
}

export function saveChartData(chartSate) {
	return dispatch => {
		dispatch(requestChartSave());
		return saveData(chartSate)
			.then(() => dispatch(completeChartSave()));
	}
}
