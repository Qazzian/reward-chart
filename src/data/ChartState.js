import * as chartActions from "./ChartActions";

const initialState = [];

export default function ChartState(state = initialState, action = {}) {
	switch (action.type) {
		case chartActions.ADD_CHART:
			return addChart(state, action.data);
		case chartActions.REMOVE_CHART:
			return removeChart(state, action.id);
		default:
			return state;
	}
}

function addChart(chartList, newChart) {
	return chartList.slice().concat(Object.assign({}, newChart));
}

function removeChart(chartList, chartId) {
	return chartList.filter((chart) => {
		return chart.id !== chartId;
	})
}
