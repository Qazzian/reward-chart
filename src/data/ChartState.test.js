import clock from 'jest-plugin-clock';

import ChartState, { chartExists } from './ChartState';
import * as ChartActions from './ChartActions';

describe('Chart State', () => {
	beforeEach(() => {
		clock.set('2018-12-06');
	});

	afterEach(() => {
	});

	it('should be defined', (done) => {
		expect(ChartState).toBeDefined();
		done();
	});

	it('should return an empty chart list by default', (done) => {
		expect(ChartState()).toMatchObject([]);
		done();
	});

	it('should add a Chart to the list', (done) => {
		const firstState = ChartState([], ChartActions.addChart({name: 123}));
		expect(firstState.length).toBe(1);
		expect(firstState[0].name).toBe(123);

		const secondState = ChartState(firstState, ChartActions.addChart({name: 987}));
		expect(secondState.length).toBe(2);
		expect(secondState[1].name).toBe(987);

		expect(firstState.length).toBe(1);
		done();
	});

	it('should ignore duplicate names', () => {
		const firstState = ChartState([], ChartActions.addChart({name: '123'}));
		expect(firstState.length).toBe(1);
		expect(firstState[0].name).toBe('123');

		const secondState = ChartState(firstState, ChartActions.addChart({name: '123'}));
		expect(secondState.length).toBe(1);
	});

	it('should Remove a chart from the list', (done) => {
		const firstState = ChartState([], ChartActions.addChart({name: 123}));
		const secondState = ChartState(firstState, ChartActions.addChart({name: 987}));
		expect(secondState.length).toBe(2);

		const thirdState = ChartState(secondState, ChartActions.removeChart({name: 987}));
		expect(thirdState.length).toBe(1);
		done();
	});

	it('should Add an emote to a chart', (done) => {
		clock.set('2018-12-14');
		const firstState = ChartState([], ChartActions.addChart({name: 123}));
		const secondState = ChartState(firstState, ChartActions.chartHappy({name: 123}));
		expect(secondState).toMatchSnapshot();
		const thirdState = ChartState(secondState, ChartActions.chartSad({name: 123}));
		expect(thirdState).toMatchSnapshot();
		done();
	});

	it('should add multiple emotes to a chart', () => {
		const firstState = ChartState([], ChartActions.addChart({name: 123}));
		const secondState = ChartState(firstState, ChartActions.chartHappy({name: 123}, new Date('2018-12-14')));
		const thirdState = ChartState(secondState, ChartActions.chartSad({name: 123}, new Date('2018-12-15')));
		const forthState = ChartState(thirdState, ChartActions.chartHappy({name: 123}, new Date('2018-12-16')));
		expect(forthState).toMatchSnapshot();
	});

	describe('Helper Functions', () => {
		test('chartExists', () => {
				const firstState = ChartState([], ChartActions.addChart({name: 123}));
				expect(chartExists(firstState, 123)).toBe(true);
		});
	});
});
