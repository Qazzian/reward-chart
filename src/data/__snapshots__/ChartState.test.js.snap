// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart State should Add a happy emote to a chart 1`] = `
Array [
  Object {
    "emotes": Array [
      Object {
        "date": "date1",
        "emote": "HAPPY",
      },
    ],
    "name": 123,
  },
]
`;

exports[`Chart State should Add a happy emote to a chart 2`] = `
Array [
  Object {
    "emotes": Array [
      Object {
        "date": "date2",
        "emote": "SAD",
      },
    ],
    "name": 123,
  },
]
`;
